universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -47,7 +71,18 @@ public ProgramServiceImpl(ProgramRepository programRepository) {
   public CompletionStage<Optional<ProgramDefinition>> getProgramDefinitionAsync(long id) {
     return programRepository
         .lookupProgram(id)
-        .thenApply(optionalProgram -> optionalProgram.map(Program::getProgramDefinition));
+        .thenComposeAsync(
+            programMaybe -> {
+              if (programMaybe.isEmpty()) {
+                return CompletableFuture.completedFuture(Optional.empty());
+              }
+              return programMaybe
+                  .map(Program::getProgramDefinition)
+                  .map(this::syncProgramDefinitionQuestions)
+                  .get()
+                  .thenApply(Optional::of);
+            },
+            httpExecutionContext.current());
```

Comment:
praise: I think this is what you mentioned in chat being a little gross. I think it's pretty alright! There's a decent amount of complexity to deal with here and this is about as readable as I feel like it can it.

Reply:
Some *might* prefer a ternary but I think it's pretty much down to personal taste at this point

```java
return programMaybe.isEmpty()
    ? CompletableFuture.completedFuture(Optional.empty())
    : programMaybe
                  .map(Program::getProgramDefinition)
                  .map(this::syncProgramDefinitionQuestions)
                  .get()
                  .thenApply(Optional::of);

```

Reply:
I'm a ternary fan, but the `if/else` is pretty clean!

Reply:
changed!